File: packages/blink-engine/src/timeline/Timeline.ts
/**
 * Timeline System
 * Manages event scheduling and timing
 */

import { EntityId } from '../ecs/Store';
import { IRFieldValue } from '../ir/types';

export interface ScheduledEvent {
  /** Unique event ID */
  id: number;
  /** Event type name */
  eventType: string;
  /** When the event should trigger (simulation time in seconds) */
  time: number;
  /** Source entity (if applicable) */
  source?: EntityId;
  /** Target entity (if applicable) */
  target?: EntityId;
  /** Additional event fields */
  fields?: Record<string, IRFieldValue>;
  /** Ordering for same-time events (lower = earlier) */
  sequence: number;
  /** If true, this event will automatically reschedule itself */
  recurring?: boolean;
  /** Interval for recurring events (in seconds) */
  interval?: number;
}

/**
 * Priority queue for scheduled events
 * Uses binary heap for O(log n) operations
 */
export class Timeline {
  private events: ScheduledEvent[] = [];
  private currentTime: number = 0;
  private nextEventId: number = 0;
  private nextSequence: number = 0;

  /**
   * Get the current simulation time
   */
  getTime(): number {
    return this.currentTime;
  }

  /**
   * Set the current simulation time
   */
  setTime(time: number): void {
    this.currentTime = time;
  }

  /**
   * Schedule an event at current time + delay
   */
  schedule(
    eventType: string,
    delay: number = 0,
    options: {
      source?: EntityId;
      target?: EntityId;
      fields?: Record<string, IRFieldValue>;
      recurring?: boolean;
      interval?: number;
    } = {}
  ): number {
    return this.scheduleAt(eventType, this.currentTime + delay, options);
  }

  /**
   * Schedule an event at an absolute time
   */
  scheduleAt(
    eventType: string,
    time: number,
    options: {
      source?: EntityId;
      target?: EntityId;
      fields?: Record<string, IRFieldValue>;
      recurring?: boolean;
      interval?: number;
    } = {}
  ): number {
    const event: ScheduledEvent = {
      id: this.nextEventId++,
      eventType,
      time: Math.max(time, this.currentTime), // Can't schedule in the past
      source: options.source,
      target: options.target,
      fields: options.fields,
      sequence: this.nextSequence++,
      recurring: options.recurring,
      interval: options.interval,
    };

    this.insert(event);
    return event.id;
  }

  /**
   * Schedule an immediate event (processed this tick)
   */
  scheduleImmediate(
    eventType: string,
    options: {
      source?: EntityId;
      target?: EntityId;
      fields?: Record<string, IRFieldValue>;
    } = {}
  ): number {
    return this.scheduleAt(eventType, this.currentTime, options);
  }

  /**
   * Schedule a recurring event
   * The event will automatically reschedule itself after each execution
   * Returns the event ID which can be used to cancel the recurring event
   */
  scheduleRecurring(
    eventType: string,
    interval: number,
    options: {
      delay?: number;
      source?: EntityId;
      target?: EntityId;
      fields?: Record<string, IRFieldValue>;
    } = {}
  ): number {
    const startTime = this.currentTime + (options.delay ?? 0);
    return this.scheduleAt(eventType, startTime, {
      source: options.source,
      target: options.target,
      fields: options.fields,
      recurring: true,
      interval,
    });
  }

  /**
   * Get the next event without removing it
   */
  peek(): ScheduledEvent | undefined {
    return this.events[0];
  }

  /**
   * Remove and return the next event
   */
  pop(): ScheduledEvent | undefined {
    if (this.events.length === 0) {
      return undefined;
    }

    const result = this.events[0];
    const last = this.events.pop()!;

    if (this.events.length > 0) {
      this.events[0] = last;
      this.siftDown(0);
    }

    // Advance time to this event
    this.currentTime = result.time;

    // If this is a recurring event, reschedule it with the SAME ID
    if (result.recurring && result.interval && result.interval > 0) {
      const nextEvent: ScheduledEvent = {
        id: result.id, // Reuse the same ID
        eventType: result.eventType,
        time: result.time + result.interval,
        source: result.source,
        target: result.target,
        fields: result.fields,
        sequence: this.nextSequence++,
        recurring: true,
        interval: result.interval,
      };
      this.insert(nextEvent);
    }

    return result;
  }

  /**
   * Check if there are any events scheduled
   */
  hasEvents(): boolean {
    return this.events.length > 0;
  }

  /**
   * Get the number of scheduled events
   */
  getEventCount(): number {
    return this.events.length;
  }

  /**
   * Cancel an event by ID
   * Returns true if the event was found and removed
   */
  cancel(eventId: number): boolean {
    const index = this.events.findIndex(e => e.id === eventId);
    if (index === -1) {
      return false;
    }

    // Remove by replacing with last element and re-heapifying
    const last = this.events.pop()!;
    if (index < this.events.length) {
      this.events[index] = last;
      this.siftDown(index);
      this.siftUp(index);
    }

    return true;
  }

  /**
   * Clear all scheduled events
   */
  clear(): void {
    this.events = [];
    this.currentTime = 0;
    this.nextSequence = 0;
  }

  /**
   * Get all events (for debugging)
   */
  getAllEvents(): ScheduledEvent[] {
    return [...this.events].sort((a, b) => {
      if (a.time !== b.time) return a.time - b.time;
      return a.sequence - b.sequence;
    });
  }

  // ===== Heap operations =====

  private insert(event: ScheduledEvent): void {
    this.events.push(event);
    this.siftUp(this.events.length - 1);
  }

  private siftUp(index: number): void {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.compare(this.events[index], this.events[parentIndex]) >= 0) {
        break;
      }
      this.swap(index, parentIndex);
      index = parentIndex;
    }
  }

  private siftDown(index: number): void {
    const length = this.events.length;
    while (true) {
      const leftChild = 2 * index + 1;
      const rightChild = 2 * index + 2;
      let smallest = index;

      if (leftChild < length && this.compare(this.events[leftChild], this.events[smallest]) < 0) {
        smallest = leftChild;
      }

      if (rightChild < length && this.compare(this.events[rightChild], this.events[smallest]) < 0) {
        smallest = rightChild;
      }

      if (smallest === index) {
        break;
      }

      this.swap(index, smallest);
      index = smallest;
    }
  }

  private compare(a: ScheduledEvent, b: ScheduledEvent): number {
    if (a.time !== b.time) {
      return a.time - b.time;
    }
    return a.sequence - b.sequence;
  }

  private swap(i: number, j: number): void {
    [this.events[i], this.events[j]] = [this.events[j], this.events[i]];
  }
}

File: packages/blink-runtime/src/timeline.rs
/// Timeline: binary heap priority queue for event scheduling.
/// Events are ordered by (time, sequence) for deterministic ordering.

use crate::event::{Event, EventId};

/// A scheduled event in the timeline.
#[derive(Clone, Debug)]
pub struct ScheduledEvent {
    pub time: f64,
    pub sequence: u64,
    pub event: Event,
}

/// Binary heap timeline for deterministic event scheduling.
pub struct Timeline {
    events: Vec<ScheduledEvent>,
    current_time: f64,
    next_sequence: u64,
    next_event_id: EventId,
    cancelled: std::collections::HashSet<EventId>,
}

impl Timeline {
    pub fn new() -> Self {
        Timeline {
            events: Vec::new(),
            current_time: 0.0,
            next_sequence: 0,
            next_event_id: 1,
            cancelled: std::collections::HashSet::new(),
        }
    }

    /// Get the current simulation time.
    pub fn get_time(&self) -> f64 {
        self.current_time
    }

    /// Set the current simulation time.
    pub fn set_time(&mut self, time: f64) {
        self.current_time = time;
    }

    /// Schedule an event at current_time + delay.
    pub fn schedule_delay(&mut self, delay: f64, mut event: Event) -> EventId {
        let event_id = self.next_event_id;
        self.next_event_id += 1;
        event.event_id = event_id;

        let scheduled = ScheduledEvent {
            time: self.current_time + delay,
            sequence: self.next_sequence,
            event,
        };
        self.next_sequence += 1;
        self.push(scheduled);
        event_id
    }

    /// Schedule an event at the current time (immediate).
    pub fn schedule_immediate(&mut self, mut event: Event) -> EventId {
        let event_id = self.next_event_id;
        self.next_event_id += 1;
        event.event_id = event_id;

        let scheduled = ScheduledEvent {
            time: self.current_time,
            sequence: self.next_sequence,
            event,
        };
        self.next_sequence += 1;
        self.push(scheduled);
        event_id
    }

    /// Schedule an event at an absolute time.
    pub fn schedule_at(&mut self, time: f64, mut event: Event) -> EventId {
        let event_id = self.next_event_id;
        self.next_event_id += 1;
        event.event_id = event_id;

        let scheduled = ScheduledEvent {
            time,
            sequence: self.next_sequence,
            event,
        };
        self.next_sequence += 1;
        self.push(scheduled);
        event_id
    }

    /// Pop the next event from the timeline. Advances current_time.
    /// Skips cancelled events.
    pub fn pop(&mut self) -> Option<Event> {
        loop {
            if self.events.is_empty() {
                return None;
            }
            let scheduled = self.pop_min();
            self.current_time = scheduled.time;

            if self.cancelled.remove(&scheduled.event.event_id) {
                continue; // Skip cancelled events
            }

            return Some(scheduled.event);
        }
    }

    /// Peek at the next event without removing it.
    pub fn peek_time(&self) -> Option<f64> {
        self.events.first().map(|e| e.time)
    }

    /// Check if there are pending events.
    pub fn has_events(&self) -> bool {
        !self.events.is_empty()
    }

    /// Cancel an event by its ID.
    pub fn cancel(&mut self, event_id: EventId) {
        self.cancelled.insert(event_id);
    }

    /// Get the number of pending events (including cancelled ones).
    pub fn len(&self) -> usize {
        self.events.len()
    }

    /// Check if the timeline is empty.
    pub fn is_empty(&self) -> bool {
        self.events.is_empty()
    }

    /// Reset the timeline to initial state.
    pub fn reset(&mut self) {
        self.events.clear();
        self.current_time = 0.0;
        self.next_sequence = 0;
        self.next_event_id = 1;
        self.cancelled.clear();
    }

    // ── Binary heap operations ──

    fn push(&mut self, event: ScheduledEvent) {
        self.events.push(event);
        self.sift_up(self.events.len() - 1);
    }

    fn pop_min(&mut self) -> ScheduledEvent {
        let last = self.events.len() - 1;
        self.events.swap(0, last);
        let min = self.events.pop().unwrap();
        if !self.events.is_empty() {
            self.sift_down(0);
        }
        min
    }

    fn sift_up(&mut self, mut idx: usize) {
        while idx > 0 {
            let parent = (idx - 1) / 2;
            if self.compare(idx, parent) {
                self.events.swap(idx, parent);
                idx = parent;
            } else {
                break;
            }
        }
    }

    fn sift_down(&mut self, mut idx: usize) {
        let len = self.events.len();
        loop {
            let left = 2 * idx + 1;
            let right = 2 * idx + 2;
            let mut smallest = idx;

            if left < len && self.compare(left, smallest) {
                smallest = left;
            }
            if right < len && self.compare(right, smallest) {
                smallest = right;
            }

            if smallest != idx {
                self.events.swap(idx, smallest);
                idx = smallest;
            } else {
                break;
            }
        }
    }

    /// Returns true if event at idx_a should come before event at idx_b.
    fn compare(&self, idx_a: usize, idx_b: usize) -> bool {
        let a = &self.events[idx_a];
        let b = &self.events[idx_b];
        if a.time != b.time {
            a.time < b.time
        } else {
            a.sequence < b.sequence
        }
    }
}

impl Default for Timeline {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::interning::StringInterner;

    fn make_event(interner: &mut StringInterner, name: &str) -> Event {
        Event::new(interner.intern(name))
    }

    #[test]
    fn test_schedule_and_pop() {
        let mut interner = StringInterner::new();
        let mut timeline = Timeline::new();

        timeline.schedule_delay(1.0, make_event(&mut interner, "A"));
        timeline.schedule_delay(0.5, make_event(&mut interner, "B"));
        timeline.schedule_delay(2.0, make_event(&mut interner, "C"));

        let e1 = timeline.pop().unwrap();
        assert_eq!(e1.event_type, interner.intern("B"));
        assert_eq!(timeline.get_time(), 0.5);

        let e2 = timeline.pop().unwrap();
        assert_eq!(e2.event_type, interner.intern("A"));
        assert_eq!(timeline.get_time(), 1.0);

        let e3 = timeline.pop().unwrap();
        assert_eq!(e3.event_type, interner.intern("C"));
        assert_eq!(timeline.get_time(), 2.0);

        assert!(timeline.pop().is_none());
    }

    #[test]
    fn test_deterministic_ordering() {
        let mut interner = StringInterner::new();
        let mut timeline = Timeline::new();

        // Schedule events at the same time - should come out in insertion order
        timeline.schedule_delay(1.0, make_event(&mut interner, "First"));
        timeline.schedule_delay(1.0, make_event(&mut interner, "Second"));
        timeline.schedule_delay(1.0, make_event(&mut interner, "Third"));

        let e1 = timeline.pop().unwrap();
        let e2 = timeline.pop().unwrap();
        let e3 = timeline.pop().unwrap();

        assert_eq!(e1.event_type, interner.intern("First"));
        assert_eq!(e2.event_type, interner.intern("Second"));
        assert_eq!(e3.event_type, interner.intern("Third"));
    }

    #[test]
    fn test_cancel_event() {
        let mut interner = StringInterner::new();
        let mut timeline = Timeline::new();

        let id1 = timeline.schedule_delay(1.0, make_event(&mut interner, "A"));
        timeline.schedule_delay(2.0, make_event(&mut interner, "B"));

        timeline.cancel(id1);

        let e = timeline.pop().unwrap();
        assert_eq!(e.event_type, interner.intern("B"));
        assert_eq!(timeline.get_time(), 2.0);
    }

    #[test]
    fn test_immediate_scheduling() {
        let mut interner = StringInterner::new();
        let mut timeline = Timeline::new();

        timeline.set_time(5.0);
        timeline.schedule_immediate(make_event(&mut interner, "Now"));

        let e = timeline.pop().unwrap();
        assert_eq!(e.event_type, interner.intern("Now"));
        assert_eq!(timeline.get_time(), 5.0);
    }

    #[test]
    fn test_has_events() {
        let mut interner = StringInterner::new();
        let mut timeline = Timeline::new();

        assert!(!timeline.has_events());
        timeline.schedule_delay(1.0, make_event(&mut interner, "A"));
        assert!(timeline.has_events());
        timeline.pop();
        assert!(!timeline.has_events());
    }

    #[test]
    fn test_reset() {
        let mut interner = StringInterner::new();
        let mut timeline = Timeline::new();

        timeline.schedule_delay(1.0, make_event(&mut interner, "A"));
        timeline.pop();
        assert_eq!(timeline.get_time(), 1.0);

        timeline.reset();
        assert_eq!(timeline.get_time(), 0.0);
        assert!(!timeline.has_events());
    }
}
